Correctly convert namedtuples in DDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44220 Correctly convert namedtuples in DDP**

Closes https://github.com/pytorch/pytorch/issues/44009
Currently if a dataloader returns objects created with a
`collections.namedtuple` or `typing.NamedTuple`, this will incorrectly be cast to a tuple. As a result, if we have data of these types, there can be runtime errors during the forward pass if the module is expecting a named tuple.

Fix this in
`scatter_gather.py` to resolve the issue reported in
https://github.com/pytorch/pytorch/issues/44009

Differential Revision: [D23536752](https://our.internmc.facebook.com/intern/diff/D23536752/)

rohan-varma committed:
Correctly convert namedtuples in DDP

Closes https://github.com/pytorch/pytorch/issues/44009
Currently if a dataloader returns objects created with a
collections.namedtuple, this will incorrectly be cast to a tuple. As a result, if we have data of these types, there can be runtime errors during the forward pass if the module is expecting a named tuple.

Fix this in
`scatter_gather.py` to resolve the issue reported in
https://github.com/pytorch/pytorch/issues/44009

Differential Revision: [D23536752](https://our.internmc.facebook.com/intern/diff/D23536752/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Correctly convert namedtuples in DDP"


Closes https://github.com/pytorch/pytorch/issues/44009
Currently if a dataloader returns objects created with a
`collections.namedtuple` or `typing.NamedTuple`, this will incorrectly be cast to a tuple. As a result, if we have data of these types, there can be runtime errors during the forward pass if the module is expecting a named tuple.

Fix this in
`scatter_gather.py` to resolve the issue reported in
https://github.com/pytorch/pytorch/issues/44009

Differential Revision: [D23536752](https://our.internmc.facebook.com/intern/diff/D23536752/)

[ghstack-poisoned]

rohan-varma commented:
Not sure who the best reviewer for this change is since `scatter_gather.py` hasn't really been touched very often, so let me know if there are better reviewers than distributed folks.

pritamdamania87 commented on test/distributed/test_distributed.py:
```diff
@@ -3130,6 +3131,47 @@ def test_broadcast_object_list(self):
         dist.broadcast_object_list(objects, src=0)
         self.assertEqual(objects, collectives_object_test_list)
 
+    @require_backend({"nccl", "gloo"})
+    @require_n_gpus_for_nccl_backend(int(os.environ["WORLD_SIZE"]), os.environ["BACKEND"])
```
Can we also add some unit test that directly test the `scatter` API with a variety of test cases? For example the case for `namedtuple(a=1, b=1)` you mentioned in the docs is a good test.

rohan-varma replied:
I added this test in `test/test_cuda.py` since it doesn't look like there were tests for this API anywhere else.

rohan-varma committed:
Update on "Correctly convert namedtuples in DDP"


Closes https://github.com/pytorch/pytorch/issues/44009
Currently if a dataloader returns objects created with a
`collections.namedtuple` or `typing.NamedTuple`, this will incorrectly be cast to a tuple. As a result, if we have data of these types, there can be runtime errors during the forward pass if the module is expecting a named tuple.

Fix this in
`scatter_gather.py` to resolve the issue reported in
https://github.com/pytorch/pytorch/issues/44009

Differential Revision: [D23536752](https://our.internmc.facebook.com/intern/diff/D23536752/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Correctly convert namedtuples in DDP"


Closes https://github.com/pytorch/pytorch/issues/44009
Currently if a dataloader returns objects created with a
`collections.namedtuple` or `typing.NamedTuple`, this will incorrectly be cast to a tuple. As a result, if we have data of these types, there can be runtime errors during the forward pass if the module is expecting a named tuple.

Fix this in
`scatter_gather.py` to resolve the issue reported in
https://github.com/pytorch/pytorch/issues/44009

Differential Revision: [D23536752](https://our.internmc.facebook.com/intern/diff/D23536752/)

[ghstack-poisoned]

rohan-varma commented:
@pritamdamania87 Added the tests for scatter, any more thoughts on this diff? Thanks!

pritamdamania87 commented on test/test_cuda.py:
```diff
@@ -3049,6 +3051,35 @@ def test_matmul_device_mismatch(self):
+
+        scatter_out = scatter_gather.scatter(inp, target_gpus)
+        for x in scatter_out:
+            self.assertTrue(isinstance(x, type(inp)))
```
Can we also assert that the `_fields` attribute is preserved correctly and also the appropriate values?

rohan-varma replied:
Thanks, adding the tests for valued actually uncovered a bug in the implementation that I just fixed.

rohan-varma committed:
Update on "Correctly convert namedtuples in DDP"


Closes https://github.com/pytorch/pytorch/issues/44009
Currently if a dataloader returns objects created with a
`collections.namedtuple` or `typing.NamedTuple`, this will incorrectly be cast to a tuple. As a result, if we have data of these types, there can be runtime errors during the forward pass if the module is expecting a named tuple.

Fix this in
`scatter_gather.py` to resolve the issue reported in
https://github.com/pytorch/pytorch/issues/44009

Differential Revision: [D23536752](https://our.internmc.facebook.com/intern/diff/D23536752/)

[ghstack-poisoned]
